Lib/test/test_typing.py

```diff
@@ -447,6 +448,321 @@ def test_bad_var_substitution(self):
+        # replacement_combo would be e.g.
+        #   [("word1", "playful"), ("word2", "feisty")]
+        templates_with_replacements = []
+        for template in templates:
+            for original, replacement in replacement_combo:
+                template = template.replace(original, replacement)
+            templates_with_replacements.append(template)
+        rendered_templates.append(tuple(templates_with_replacements))
+
+    # Finally, remove the duplicates (but keep the order).
+    rendered_templates_no_duplicates = []
+    for x in rendered_templates:
+        # Inefficient, but should be fine for our purposes.
+        if x not in rendered_templates_no_duplicates:
+            rendered_templates_no_duplicates.append(x)
+
+    return rendered_templates_no_duplicates
+
+
+class TemplateReplacementTests(BaseTestCase):
+
+    def test_two_templates_two_replacements_yields_correct_renders(self):
+        actual = template_replace(
+                templates=["Cats are word1", "Dogs are word2"],
+                replacements={
+                    "word1": ["small", "cute"],
+                    "word2": ["big", "fluffy"],
+                },
+        )
+        expected = [
+            ("Cats are small", "Dogs are big"),
+            ("Cats are small", "Dogs are fluffy"),
+            ("Cats are cute", "Dogs are big"),
+            ("Cats are cute", "Dogs are fluffy"),
+        ]
+        self.assertEqual(actual, expected)
+
+    def test_no_duplicates_if_replacement_not_in_templates(self):
+        actual = template_replace(
+                templates=["Cats are word1", "Dogs!"],
+                replacements={
+                    "word1": ["small", "cute"],
+                    "word2": ["big", "fluffy"],
+                },
+        )
+        expected = [
+            ("Cats are small", "Dogs!"),
+            ("Cats are cute", "Dogs!"),
+        ]
+        self.assertEqual(actual, expected)
+
+
+class GenericAliasSubstitutionTests(BaseTestCase):
+
+    def test_one_parameter(self):
+        T = TypeVar('T')
+        Ts = TypeVarTuple('Ts')
+
+        class C(Generic[T]): pass
+
+        generics = ['C', 'list', 'List']
+        tuple_types = ['tuple', 'Tuple']
+
+        tests = [
+            # Alias                               # Args                     # Expected result
+            ('generic[T]',                        '[()]',                    'TypeError'),
+            ('generic[T]',                        '[int]',                   'generic[int]'),
+            ('generic[T]',                        '[int, str]',              'TypeError'),
+            ('generic[T]',                        '[tuple_type[int, ...]]',  'generic[tuple_type[int, ...]]'),
+            ('generic[T]',                        '[*tuple_type[()]]',       'generic[*tuple_type[()]]'),
+            ('generic[T]',                        '[*tuple_type[int]]',      'generic[*tuple_type[int]]'),
```

Comment:
Ditto above comment - we'd be ok with outlawing e.g. `list[*tuple[int]]`?

Reply:
My instinct is to disallow as little as possible at runtime, but I struggle to see why you'd ever want to do something like `list[*tuple[int]]`.

Reply:
> My instinct is to disallow as little as possible at runtime, but I struggle to see why you'd ever want to do something like list[*tuple[int]].

(Agreed that it seems like an unlikely thing to want to do.)

Reply:
I agree with Jelle that we should disallow as little as possible at runtime. The only reason I bring this up is because there seem to be inconsistencies in the (current) runtime implementation, such as: 

```
            ('generic[T]',                        '[*tuple_type[int]]',      'generic[*tuple_type[int]]'),
            ('generic[T1, T2]',                        '[*tuple_type[int, str]]',                           'TypeError'),  # Should be generic[*tuple_type[int, str]]
```

Either allow everything at runtime or forbid all `*` at runtime. Doing it half-way just seems wrong and confusing from a user's point of view.

Reply:
Fixed.